### ISSUE TITLE
Feature request: convert str(Path) to fsdecode(Path)

### ISSUE THREAD

**Issue by m-aciek** [charliermarsh/ruff#3675]
I [learned today](https://discuss.python.org/t/make-pathlib-extensible/3428/79) that [os.fsdecode(Path)](https://docs.python.org/3/library/os.html#os.fsdecode) is preferred over str(Path) for [pathlib.Path](https://docs.python.org/3/library/pathlib.html) objects. Ruff could have a check for str(Path) usages.

**Comment by nstarman** [comment]
Isn't `.as_posix()` preferred over both?

**Comment by edgarrmondragon** [comment]
I think this is hard to accomplish in a useful way without type checking. The only cases this would catch would be when the `Path` object is directly instantiated inside the `str()` call.

**Comment by m-aciek** [comment]
Hm, `fspath()` is runtime utility to ensure that program really operates on a path, this way it does more than type checking – type checking doesn't prevent always from having a not expected type in runtime. But I agree that for many (majority of?) cases, when path object doesn't come as arbitrary object from outside (API?), the `fspath()` usage wouldn't be needed. Probably recommending using `fspath()` for all cases would be counter-productive.

@brettcannon do you maybe have an opinion on it?

**Comment by brettcannon** [comment]
Since I was asked for my opinion, here's my opinionated opinion 😉:

1. Never use `str(Path)`; can give you `None`
2. Only use `PurePath.as_posix()` if you specifically need the path in a POSIX format
3. Use `os.fsdecode(path)` if you need the string encoding of the `Path` object (don't need to use `os.fspath()` since `pathlib` doesn't deal in `bytes`), but you should avoid this as much as possible and only do it last-minute when you must
4. To be fully generic/duck-typing, us `os.fspath()` anywhere you accept a path and don't care if it's a `str` or `bytes`, else use `os.fsdecode()` or `os.fsencode()`, respectively
5. To make this work you will need type hints

**Comment by ericrbg-harmonic** [comment]
My understanding is that now we should be using `os.fspath` instead of `fsdecode`, fwiw

**Comment by brettcannon** [comment]
> My understanding is that now we should be using `os.fspath` instead of `fsdecode`, fwiw

Who's "we" in that sentence?

**Comment by ericrbg-harmonic** [comment]
Apologies, "we" was meant to be interpreted as "one", as in "one should", although a closer reading of [PEP 519](https://peps.python.org/pep-0519/) suggests that both have value (and possibly `fsdecode` is more specific for the string case?)

E: and I missed your above comments, Brett - sorry 🤦

**Comment by brettcannon** [comment]
> `fsdecode` is more specific for the string case?

Correct. When I wrote PEP 519 approaching 10 years ago, there was still a need/desire to support paths in binary format on Unix OSs, but UTF-8 seems to have won that battle, so it isn't really a case I think much code is set up to handle.
